### DISCUSSION TITLE
Provenance doesn’t show up

### DISCUSSION THREAD

**Discussion by remcohaszing** [npm/feedback#910] [General]
I just created a new package: [`remark-preset-remcohaszing`](https://github.com/remcohaszing/remark-preset-remcohaszing).

1. I enabled provenance via [`.npmrc`](https://github.com/remcohaszing/remark-preset-remcohaszing/blob/v1.0.1/.npmrc).
2. The package is built using the [`pack`](https://github.com/remcohaszing/remark-preset-remcohaszing/blob/v1.0.1/.github/workflows/ci.yaml#L20-L32) job.
3. The package is published using the [`release`](https://github.com/remcohaszing/remark-preset-remcohaszing/blob/v1.0.1/.github/workflows/ci.yaml#L64-L80) job.
4. The logs of the workflow can be seen [here](https://github.com/remcohaszing/remark-preset-remcohaszing/actions/runs/4756035302). They show npm 9.6.4 is used, which I believe should support provenance, but no related CI output.

However, on https://www.npmjs.com/package/remark-preset-remcohaszing I don’t see any provenance information.

Am I missing something? Is there a delay maybe?

**Comment by ljharb** [comment]
Can you _get_ provenance when publishing a previously created tarball?

**Reply by remcohaszing** [reply]
Is that question targeted at me or the npm team? Because I don't know.

**Reply by ljharb** [reply]
Both I guess :-)

**Reply by remcohaszing** [reply]
FWIW I added `ls -lah` as a step to the `pack` job, and that doesn”t generate a provenance file. https://github.com/remcohaszing/remark-preset-remcohaszing/actions/runs/4758353045/jobs/8456302208#step:6:1

**Reply by ljharb** [reply]
I wouldn't expect it to, only on the publish step.

**Reply by feelepxyz** [reply]
Yeah this should work, we're really binding where the publish came from. Having access to build logs should also help if you want to understand what steps where involved in generating the tarball after the fact. You could put anything in there if you really wanted to, similar to running other build scripts before running `npm pack/publish`. What you couldn't do is spoof which repo the publish originated from.

**Reply by ljharb** [reply]
ok, so provenance isn’t about the files at all, it’s only about where the publish happened?

**Reply by feelepxyz** [reply]
Yes in the sense that we can only get authenticated source details about where the workflow/CI job ran, hence were the `npm publish` happens. This gives you a verifiable link to the build logs and particular job where the publish happens and logs for determining what happened before. We're also planning to allow maintainers to "pin" a particular source repo as the one that is allowed to publish, so it can't be easily changed/removed between versions.

This obviously still leaves some holes, as a sneaky attacker could clone a completely different repo and run some sneaky build scripts fetched from the internet that return different results depending in when they are fetched.

The plan is to also offer [reusable workflows](https://docs.github.com/en/actions/using-workflows/reusing-workflows) (or [Orbs in Circle](https://circleci.com/orbs/)) for publishing npm packages that prevent the calling workflow from making modifications to the publish flow, locking this down considerably, but also makes it hard to support the wide range of custom workflows that exist today. 

There would still be the problem of running arbitrary build scripts but hopefully in future we'll have the ability to run these builds in a job/workflow step that doesn't have internet access.

**Reply by ljharb** [reply]
@feelepxyz that's very surprising given all of the marketing around "provenance" - i'm not sure how the location of the publish matters to security given that the contents aren't actually verifiable at all.

Additionally, github actions erases logs after 90 days (by default, i think?) - so it's not much of a permanently auditable context.

**Reply by feelepxyz** [reply]
Agreed, logs don't give a permanently auditable context but better than nothing.

Provenance is a first step of many that will hopefully over time build up to provide significant protections against supply chain attacks. We are very early in this journey!

> given that the contents aren't actually verifiable at all.

While this is true, I think it's helpful to break this down into different attack vectors. We're explicitly not dealing with a bunch of attacks that we need to get to eventually.

Provenance as it exists today does not protect against **source repository** or **build system compromises in** any way. What you gain is a way to associate trust with a given source repository and build instructions, if you assume they haven't been compromised. This is obviously not complete and does not scale per package and repository. I think it's still a step in the right direction as it allows us to start trusting for example, that `react` on npmjs.com comes from `github.com/facebook/react` and I have some trust that the files in this repo have not been compromised or that the build instructions have not been tampered with. And in trust, I mean relatively to something else like `github.com/randouser/react`. Official repos get more eyeballs so issues are hopefully spotted earlier at the very least.

While provenance does not ensure the contents of the package are verifiable, or reproducible from the exact source commit, it does give you a verifiable link to the event that triggered a given workflow at a specific commit. It's then up to you to decide if you trust the what's in the source code. 

This doesn't really scale and we can do better! Next steps are figuring out how we can protect against build compromises and source code compromises and scale this. So even if someone compromises your source code, they can't sneak something in via messing with the build. This could be achieved using external, reusable workflows that build the package with the exact input from the triggering workflow event and don't allow the source maintainer to make any modifications to the build.

Figuring out how to make these build instructions actually tamper-proof is going to be really hard as there are so many ways to build npm packages with a lot of dynamic stuff going on. But it might be feasible for organisations or individual maintainers like yourself to decide on a safe subset of functionality for building your packages, publishing this as a reusable actions workflow and use that to publish all your packages. Consumers could then verify a particular workflow was used to publish a package. The registry could also make this "sticky" so any subsequent publishes that don't come from a particular workflow would require 2FA to change.

Layering on this, I would love to do what you were suggesting in the [provenance RFC](https://github.com/npm/rfcs/pull/626): Re-build the package in a completely isolated build environment and determine if it's reproducible and layer this on as additional proof the build hasn't been compromised. 

Source code compromises is the other attack vector were things like branch protections and PR review help but there's also a bunch more to do.

**Reply by ljharb** [reply]
couldnt that be verified more simply and more universally by ensuring that the publishing npm user has write+ access to the repo listed?

**Comment by feelepxyz** [comment]
@remcohaszing it's not enabled by default when publishing with Actions so you'll need to add `--provenance` to the publish command: https://github.com/remcohaszing/remark-preset-remcohaszing/pull/1

**Reply by remcohaszing** [reply]
Thanks! That worked!

This is what happened: The `publish` job doesn’t clone the project, meaning `.npmrc` isn’t available. Otherwise setting `provenance = true` in `.npmrc` should have worked as an alternative to using the `--provenance` flag.

I wonder if I still need the `contents: read` permission for the release job if it doesn’t even clone the repo.

**Reply by feelepxyz** [reply]
> I wonder if I still need the `contents: read` permission for the release job if it doesn’t even clone the repo.

Probably not needed!

**Reply by remcohaszing** [reply]
I confirmed it’s not needed.
